### ISSUE TITLE
Our JS code delay page loading and use a lot of CPU (both Servo & Gecko)

### ISSUE THREAD

**Issue by paulrouget** [browserhtml/browserhtml#935]
See https://github.com/servo/servo/issues/10000

Looking at the profiler, we spend a lot of time executing JS code. CPU ends up spiking at almost 200%.
Everything is back to order when disabling the progressbar code.

Part of the overhead is due to a Servo bug we will be looking at next, but it's only a small part of the issue. Apparently, we also have a JS overhead with Gecko.

Can we spend some time investigating what's going on in our JS code?

For example, could we replace the progressbar code with a naive non-vnode-based algorithm see what happens?

**Comment by paulrouget** [comment]
I'll try to come up with some more meaningfull numbers later today.

**Comment by paulrouget** [comment]
During page load, 30% of the time is spent in the rAF loop.

**Comment by paulrouget** [comment]
> For example, could we replace the progressbar code with a naive non-vnode-based algorithm see what happens?

Just did that.

Here are my conclusions:
- 2 servo issues: rAF overhead (https://github.com/servo/servo/issues/9844) and poor style.transform performance (https://github.com/servo/servo/issues/10050) 
- we still spend a significant amount of time (~20%) in JS not doing any DOM changes

Replacing rAF with `setTimeout` and commenting out the transform change brings us down to 20%.

**Comment by paulrouget** [comment]
Here is the result of profiling browerhtml loading a page with a download latency. 85% of the samples are from the script thread: http://people.mozilla.org/~prouget/bugs/issue935.svg (the box are clickable).

**Comment by paulrouget** [comment]
In the previous comment, I used work around to avoid the 2 servo bugs mentioned above.

**Comment by paulrouget** [comment]
Note for self - need sampling for:
- browserevents no called
- browerevents called, progress.view not called
- browserevents called, progress.view called, no change transform (but changed vdom)
- with and without a fake requestAnimationFrame
- with empty slow page, with firefox.com

**Comment by paulrouget** [comment]
I gathered some more data. Nothing really new here, just solidifying the previous results.

tl;dr: as of now, loading firefox.com takes 90% of the CPU, when the actual loading takes only 5%. The 85% are used for the progressbar. We can go down to ~25% total.

4 different version of browserhtml:
* *A* normal version
* *B* browseerevents not called
* *C* progress.view not called
* *D* instead of change `progressbar.style.transform`, change `progressbar.title`

2 different websites:
* `http://firefox.com`
* `http://localhost/empty.html`,  a local http server that takes 15 seconds to reply

2 scheduling strategies:
* `requestAnimationFrame`
* `setTimeout(16)`

All of these setups tried twice.

I use this powermetrics configuration: `powermetrics --samplers tasks --show-process-gpu -n 1 -i 20000`.
See https://developer.mozilla.org/en-US/docs/Mozilla/Performance/powermetrics

Raw data:

```
code, CPU ms/s #1  ,  CPU ms/s #2  , Energy Impact  #2, Energy Impact #2,RAF,site
A,900.58,970.08,246.21,250.02,TRUE,firefox.com
A,682.72,761.19,191.56,200.01,TRUE,empty.html
A,693.27,752.2,217.37,229.96,FALSE,firefox.com
A,475.65,506.64,157.97,169.93,FALSE,empty.html
B,43.63,45.56,7.61,8.81,TRUE,firefox.com
B,16.39,11.19,3.59,3.31,TRUE,empty.html
B,48.6,43.32,8.07,7.78,FALSE,firefox.com
B,9.84,9.99,2.47,2.29,FALSE,empty.html
B,631.01,621.12,214.73,215.57,TRUE,firefox.com
C,445.68,433.31,164.59,163.32,TRUE,empty.html
C,218.49,212.29,25.09,25.58,FALSE,firefox.com
C,140.94,139.9,15.99,15.58,FALSE,empty.html
D,598.43,579.99,212.08,211.15,TRUE,firefox.com
D,460.82,463.37,167.29,167.72,TRUE,empty.html
D,233.87,218.8,27.15,24.48,FALSE,firefox.com
D,148.25,149.99,16.74,17.71,FALSE,empty.html
```

<img width="613" alt="data" src="https://cloud.githubusercontent.com/assets/373579/14274943/959b03de-fb14-11e5-8594-f0a654a02da5.png">

I've been toying with excel, and here are some conclusions:

- without the progressbar at all, bhtml takes *1-5%* of the CPU for 20 seconds. With the progressbar, it takes *70%-90%* for 20 seconds. This is not acceptable;
- complexity of the page is not relevant;
- D is barely more expensive than C;
- `requestAnimationFrame` adds a pretty big overhead. *~200-300ms/s*;
- cost of drawing (*D to A* without *rAF*) is big. *~300ms-400ms/s*;
- JS cost around *~200ms/s*;

Short term, I hope we can fix the rAF issue (servo/servo#9844) but we won't solve the drawing issue (servo/servo#10050) just yet, but I'm pretty sure we can go down to *~200ms/s* (that is what a standalone reproduction of the progressbar animation takes).

For javascript, we really don't want to spend *200ms/s* just to compute the new position of the bar.

**Comment by paulrouget** [comment]
> For javascript, we really don't want to spend 200ms/s just to compute the new position of the bar.

So the goal of this issue is to get this number down to something reasonable (a lot closer to `0`).

Maybe we should not use virtual dom for animations and use a naive animation loop. I don't know how that would play with our internal mechanism though.

@Gozala what do you think?

**Comment by metajack** [comment]
cc @pcwalton

**Comment by paulrouget** [comment]
@pcwalton is aware of the 2 servo issues mentioned above. This bug is about browserhtml JS code.

**Comment by paulrouget** [comment]
For the record, Gecko's numbers:

```
Name                               CPU ms/s  User%  Energy Impact
graphene                           439.44    89.02  113.92      
plugin-container                   108.11    65.96  10.75       
```

**Comment by Gozala** [comment]
> Maybe we should not use virtual dom for animations and use a naive animation loop. I don't know how that would play with our internal mechanism though.
> 
> @Gozala what do you think?

1. I would like to get to the bottom of it & find out what are the bottlenecks before deciding that animations should be done differently, even if we decide so I'd like to make an educated decision rather than give up.

2. Yes we can animate things say "progress bar" outside of the main update / view loop. We could just schedule a task from update function that will run separate RAF loop just to animate progress-bar. That way virtual-dom generation + diffing will be skipped (which is presumably what JS is busy with now).

**Comment by Gozala** [comment]
> > I would like to get to the bottom of it & find out what are the bottlenecks before deciding that animations should be done differently, even if we decide so I'd like to make an educated decision rather than give up.

> Any update on that?

No I'm afraid I was busy with other stuff.

> Also - we mentioned that we could have a temporary fast code path for the progressbar. Is it still planned?

https://github.com/browserhtml/browserhtml/pull/995

**Comment by paulrouget** [comment]
#995 improves the situation but it's not switch on by default. We still need to understand why it is slow in the first place as this also affect the sidebar animation.

**Comment by paulrouget** [comment]
Recent work greatly improved the situation. Closing.
